ext/objspace/objspace_dump.c

```diff
@@ -25,32 +25,135 @@
 static VALUE sym_output, sym_stdout, sym_string, sym_file;
 static VALUE sym_full;
 
+#define BUFFER_CAPACITY 4096
+
 struct dump_config {
     VALUE type;
-    FILE *stream;
+    VALUE stream;
     VALUE string;
     const char *root_category;
     VALUE cur_obj;
     VALUE cur_obj_klass;
     size_t cur_obj_references;
     unsigned int roots: 1;
     unsigned int full_heap: 1;
+    unsigned long buffer_len;
+    char buffer[BUFFER_CAPACITY];
 };
 
-PRINTF_ARGS(static void dump_append(struct dump_config *, const char *, ...), 2, 3);
 static void
-dump_append(struct dump_config *dc, const char *format, ...)
+dump_flush(struct dump_config *dc)
 {
-    va_list vl;
-    va_start(vl, format);
+    if (dc->buffer_len) {
+        if (dc->stream) {
+            size_t written = rb_io_bufwrite(dc->stream, dc->buffer, dc->buffer_len);
+            if (written < dc->buffer_len) {
+                MEMMOVE(dc->buffer, dc->buffer + written, char, dc->buffer_len - written);
+                dc->buffer_len -= written;
+                return;
+            }
+        }
+        else if (dc->string) {
+            rb_str_cat(dc->string, dc->buffer, dc->buffer_len);
+        }
+        dc->buffer_len = 0;
+    }
+}
 
-    if (dc->stream) {
-	vfprintf(dc->stream, format, vl);
+static inline void
+buffer_ensure_capa(struct dump_config *dc, unsigned long requested)
+{
+    RUBY_ASSERT(requested <= BUFFER_CAPACITY);
+    if (requested + dc->buffer_len >= BUFFER_CAPACITY) {
+        dump_flush(dc);
+        if (requested + dc->buffer_len >= BUFFER_CAPACITY) {
+            // TODO: raise?
+        }
     }
-    else if (dc->string)
-	rb_str_vcatf(dc->string, format, vl);
+}
+
+static void buffer_append(struct dump_config *dc, const char *cstr, unsigned long len)
+{
+    if (len > 0) {
+        buffer_ensure_capa(dc, len);
+        MEMCPY(dc->buffer + dc->buffer_len, cstr, char, len);
+        dc->buffer_len += len;
+    }
+}
 
-    va_end(vl);
+# define dump_append(dc, str) buffer_append(dc, (str), (long)strlen(str))
+
+static void
+dump_append_ld(struct dump_config *dc, const long number)
+{
+    buffer_ensure_capa(dc, 11); // largest long representation ("-2147483648")
+    unsigned long required = snprintf(dc->buffer + dc->buffer_len, 11, "%ld", number);
+    RUBY_ASSERT(required <= 11);
```

Comment:
Why only 11, 32bit environment?
And it isn't a good idea to spread magic numbers.

Using "ruby/util.h":
```suggestion
    const int width = DECIMAL_SIZE_OF_BITS(sizeof(number) * CHAR_BIT - 1) + 1;
    buffer_ensure_capa(dc, width);
    unsigned long required = snprintf(dc->buffer + dc->buffer_len, width, "%ld", number);
    RUBY_ASSERT(required <= width);
```

Reply:
TIL about `DECIMAL_SIZE_OF_BITS`, thanks a lot I'll integrate that.

Reply:
I forgot the room for terminator.
`width` needs +1.

Reply:
Done.